### PR TITLE
zerotier: add support for allowDefault and others

### PR DESCRIPTION
**Pull request by generalinterest**
Hi,

This is the the code I added to /etc/init.d/zerotier.  it’s just adding variables, performing a config_get for them.  
The clever bit is in add_join, where it simply writes these var’s to the network’s local.conf if the vars are present.   This is safe because any preexisting local.conf is already deleted!

### CODE REVIEW

File: zerotier/files/etc/init.d/zerotier [hunk 1]
```diff
@@ -87,6 +91,20 @@ start_instance() {
 	add_join() {
 		# an (empty) config file will cause ZT to join a network
 		touch $path/networks.d/$1.conf
+
+		# create a corresponding .local.conf file if any of it's settings exist.
+                if [ $allow_managed ]; then
```
Comment (mwarning) [inline zerotier/files/etc/init.d/zerotier:96]: Let use e.g.:
```
if [ -n "$allow_managed" ]; then
````
Otherwise we insert code into the script that it tries to execute...

Comment (generalinterest) [reply zerotier/files/etc/init.d/zerotier:96]: Hi ya,

Thank you for reviewing.  That was my first try - to keep the code consistent.

I am not sure why, but with that syntax, if "allowManaged" is not present in /etc/config/zerotier, you get a malformed line written to the .local.conf !

/etc/config/zerotier

```
config zerotier 'holidays' 
	option enabled '1'
	option allowDefault '1'
	option allowDNS '1'
	list join 'abcdef'
	option secret 'XXX'
```

Then allowManaged shows up with no value!

/var/lib/zerotier-one/networks.d/56374ac9a4ed101a.local.conf
```
allowManaged=
allowDefault=1
allowDNS=1
```

The only safe way I found was without the "-n".

Comment (mwarning) [reply zerotier/files/etc/init.d/zerotier:96]: Oh, you have put your code in the add_join() function. Do really need to do that?

Comment (ogarcia) [reply zerotier/files/etc/init.d/zerotier:96]: I think the problem lies in the way the configuration file is set up. Maybe initially the configuration as it is was fine since there were only a couple of options to configure and it could make sense to raise more than one instance of zerotier.

Right now I don't know to what extent it may be necessary to have more than one instance of zerotier running (also, as it is set up I don't think it would work well because you can only have one `/var/lib/zerotier-one`). It seems to me something very exotic that I don't think anyone has since the normal thing is to start a single zerotier and configure it with all the options that you really need. In that sense it makes more sense that the configuration is divided into a global section and then into different sections for each of the networks.

What I would do would be to have a configuration like this:
```config
config zerotier 'global'
	# enable or disable zerotier
	option enabled 0
	# persistent configuration folder (for ZT controller mode)
	#option config_path '/etc/zerotier'
	# copy <config_path> to RAM to prevent writing to flash (for ZT controller mode)
	#option copy_config_path '1'
	#option port '9993'
	# path to the local.conf
	#option local_conf '/etc/zerotier.conf'
	# Generate secret on first start
	option secret ''

config network 'mynet' # Network names are optional, for us this is `zerotier.@network[0]`
	# network id
	option id 'thenetid'
	option allow_managed '0'
	option allow_global '1'
	option allow_default '1'
	option allow_dns '0'

config network 'othernet'
	# network id
	option id 'othernetid'
	option allow_managed '1'
	option allow_global '1'
	option allow_default '1'
	option allow_dns '1'
```

@mwarning what do you think?

Comment (ogarcia) [reply zerotier/files/etc/init.d/zerotier:96]: @mwarning I haven't commented it out, but if it's OK with you I can try to rewrite `/etc/init.d/zerotier` to fit this configuration. :wink:

Comment (ogarcia) [reply zerotier/files/etc/init.d/zerotier:96]: @mwarning to make it clearer (and because it didn't give me that much work) I have put an example of how it would look in #120

I think it is a good improvement and simplifies the configuration a lot.

Comment (generalinterest) [reply zerotier/files/etc/init.d/zerotier:96]: new add_join() defensive code.  

`         if [ $allow_managed ] && [ -n $allow_managed ]; then`

means if the config line exists and also has a value, then add the entry to local.conf.

```
        add_join() {
                # an (empty) config file will cause ZT to join a network
                touch $path/networks.d/$1.conf

                # create a corresponding .local.conf file if any of it's settings exist.
                if [ $allow_managed ] && [ -n $allow_managed ]; then
                        echo "allowManaged=$allow_managed" >> $path/networks.d/$1.local.conf
                fi
                if [ $allow_global ] && [ -n allow_managed ]; then
                        echo "allowGlobal=$allow_global" >> $path/networks.d/$1.local.conf
                fi
                if [ $allow_default ] && [ allow_default ]; then
                        echo "allowDefault=$allow_default" >> $path/networks.d/$1.local.conf
                fi
                if [ $allow_dns ] && [ -n allowd_dns ]; then
                        echo "allowDNS=$allow_dns" >> $path/networks.d/$1.local.conf
                fi
        }

        config_list_foreach $cfg 'join' add_join

```

`config_list_foreach $cfg 'join' add_join`

means that it allows for multiple networks to be managed by the same zerotier-one instance.  By placing these “allow” settings in add_join() function, these vars get the same network name with just a different file suffix.  You would have to repeat the “config_list_foreach” with a different function to pull that logic out of add_join().

Comment (mwarning) [reply zerotier/files/etc/init.d/zerotier:96]: @ogarcia thank you for all your responses, but I have to read it tomorrow. :-)

Comment (generalinterest) [reply zerotier/files/etc/init.d/zerotier:96]: Hi @ogarcia 

All the “allow” variables here are optional.  Zerotier will work with it’s defaults for any missing value in .local.conf.

I believe this allows it to keep compatibility with existing /etc/config/zerotier.

I suspect these “allow’s” are chosen with care. For example, it would not make much sense to accept a default route from two different zerotier networks at the same time!

With that, I would recommend any new Global section be considered optional.  That would also ensure compatibility with prior versions then as well.

Comment (ogarcia) [reply zerotier/files/etc/init.d/zerotier:96]: @generalinterest Maintaining backward compatibility in this change is very costly and will never end up working properly. IMHO the best thing to do is not to try. What will happen if a user has the old configuration and tries to boot ZeroTier is that he will simply get a message saying `disabled in /etc/config/zerotier` and the ZeroTier daemon will not start.

What is clear is that if we want to include this functionality we have to change the way ZeroTier is currently configured because if we leave it as the proposal you make a user who is joining two networks will not be able to configure one of them as default (when he/she set `allow_default 1` will be for all networks).

TL/DR: If we want to add this functionality we would have to modify the configuration file to [something like to #120](https://github.com/mwarning/zerotier-openwrt/blob/7e04264f4a71d079497af9eb953be8f759165f38/zerotier/files/etc/config/zerotier).

Comment (generalinterest) [reply zerotier/files/etc/init.d/zerotier]: Hi @ogarcia 

Just in defense of this proposal, the “allow” vars are all zerotier network specific and not global values   It amounts to the corresponding zerotier-cli command..

`zerotier-cli set abcdef allowDefault=1`

So for sure the user would need to have an understanding what these vars mean and do!

For example, a zerotier-openwrt router (A) could be the destination for the default route on a zerotier network. This router does not need to set the default route, it is the default route for that network.  
   Then a second, remote zerotier-openwrt router (B) would set allowDefault so that devices behind that will be routed across the zerotier network to the first Router - A.  
   It would make sense that only the remote’s would need to consider allowDNS as well.

The first router - A, could be joined to additional zerotier networks, where the same process would apply.  It is only the remote clients need the option to allowDefault.

In zerotier-openwrt terms this would be a “config network” block where a router A would have a list of joins with no allowDefault.   While on router B, the “config network” block might have a single join where it does allowDefault.

In more complex setups, a zerotier-openwrt router could be the default route for some zerotier networks but set a default network from another zerotier network if it choses to!  It would require a second “config network” that specifies the options it wants.

I believe this proposal supports all that and is compatible with prior releases - no change unless you make the change!  😃

Comment (ogarcia) [reply zerotier/files/etc/init.d/zerotier]: @generalinterest I think you are not understanding the problem of your proposal, I will give you an example to see it more clearly. Imagine you have this configuration file:
```uci
config zerotier 'myconfig' 
	option enabled '1'
	option allowDefault '1'
	option allowDNS '1'
	list join 'firstnetwork'
	list join 'secondnetwork'
	option secret 'XXX'
```

This only starts one instance of ZeroTier, but configures the connection to two different networks `firstnetwork` and `secondnetwork` (which is not uncommon, I myself am connected to two ZT networks). With this configuration on both networks you are enabling a `allowDefault` because you have no way (as the configuration file is now created) to tell it that you want the `allowDefault` on `firstnetwork` and not on `secondnetwork` (to give an example). The same goes for `allowDNS`.

This has nothing to do with having something like this (which I think is what you're thinking):
```uci
config zerotier 'myconfigone' 
	option enabled '1'
	option allowDefault '1'
	option allowDNS '1'
	list join 'secondnetwork'
	option secret 'XXX'

config zerotier 'myconfigtwo' 
	option enabled '1'
	option allowDefault '0'
	option allowDNS '0'
	list join 'secondnetwork'
	option secret 'XXX'
```

The latter simply does not work. It will not start two separate instances of ZeroTier, what it will do is simply start the second `myconfigtwo` since ZeroTier can only have one configuration in `/var/lib/zerotier`. If you wanted to start the first one you would necessarily have to disable the second one (or better yet comment it out).

This is why I say that the best thing to do is to consider changing the way the configuration file is set up. Because right now as it is it does not allow different definitions per network and the functionality of configuring two or more ZeroTier instances does not make sense because the `zeroter-one` binary is going to look for the configuration to `/var/lib/zerotier` in a _forced way_ and it is not something that can be configured (in runtime I mean).

Comment (generalinterest) [reply zerotier/files/etc/init.d/zerotier]: Hi @ogarcia 

admittedly I am new to zerotier-openwrt, so I apologize if I am overstepping.  The two config's from your second example above will create two instances of zerotier-one as per the code...

```

start_service() {
        config_load 'zerotier'
        config_foreach start_instance 'zerotier'
}
```

Where start_instance includes ...
```
        procd_open_instance
        procd_set_param command $PROG $args $path
        procd_set_param stderr 1
        procd_set_param respawn
        procd_close_instance
```
This is what starts up an instance of zerotier-one for each "config zerotier".

Then for your first example, that is a case where it would be incorrect to have allowDefault and allowDNS.  For the router hosting two or more networks, it should be acting as the gateway IMHO for both networks.  Any other zerotier clients on these networks should pickup the default route to get to your router.

If there is a reason to have two networks, then perhaps adding regular routes, just not the default route, might be more appropriate.

Again I am sorry if I am pushing too hard on this.

Comment (ogarcia) [reply zerotier/files/etc/init.d/zerotier]: @generalinterest well, maybe I _cheated_ a little in my explanation, it is true that you can launch zerotier indicating a home directory and supposedly should not attack `/var/lib/zerotier` but the directory that you are indicating, just as it is right now the `/etc/init.d/zerotier` should raise more than one instance, but the reality is that whenever I have tried it has given me problems to have more than one `zerotier-one` running.

That's why I say that it makes no sense to me that you can run more than one instance, because it is problematic (or at least it was at all times when I tried it).

The summary is that if running more than one instance does not give good results (and therefore the logical thing is to run a single instance and with it join several networks) it would be best if the configuration allows you to configure these networks separately, which you cannot do with a list.

Comment (ogarcia) [reply zerotier/files/etc/init.d/zerotier]: > I take your point, in resource constrained environments, running two configs could be impossible. In my mind it comes down to the network topologies zerotier-openwrt is expected to support.

That is the question. In OpenWrt you should always think about the configuration that consumes less resources. In this particular case starting more than one ZeroTier daemon does not make sense because it does not offer any advantage and is an unnecessary consumption of resources.

> I am curious as to the use cases for there to be two or more networks on a zertotier-openwrt. In my mind, this is because they are acting as the gateway (default route) for the zerotier networks. If that is not the case, neither of them should be a default route, but avail of zerotier configured routes instead if need be.

Having two or more ZT networks is not uncommon. For example, in a concrete case, you can create one network for you and your friends (to share things), another one for your family (with the same purpose) and another one for work. They are three isolated networks that do not see each other but you may want to connect to all three from your router because you do use those three networks.

> If there is a requirement to support two or more networks where only one of them should be selected for it’s default route, then I think that would require further analysis.

Not only in that case. There are four configuration parameters and any one of them you may want on one network and not on another. For example in my previous example the network that you share with your friends is not managed by you and has established certain routes to public IPs that do not interest you, but in the work network are published certain routes to public IPs that you need (because to reach those IPs you have to go out to the Internet necessarily through your work network), in that case you will want to disable the `allowGlobal` in your friends network and enable it in the work network.

> I just did a quick test and while this might technically work, it's not pretty.
> Change the config from allowDefault to network_AllowDefault. You would only need that for the one zerotier network that needs the default route setting.

It is an option. Personally I don't like it because instead of having a standard name like `option allow_managed '1'` you will have to create a name for each network `option ztnetworname_allow_managed '1'` which can lead to many mistakes.

> I'm not sure where to go from here. All suggestions are welcome!

There is not much more we can do since the UCI system does not allow nested lists. :disappointed: 

IMHO, the best thing to do is to replace the current multi-instance model (which is not used) with the one-instance, multi-network model, but let's wait and see what @mwarning says.

Comment (generalinterest) [reply zerotier/files/etc/init.d/zerotier]: Hi @ogarcia  thank you for that detailed explanation.

I have one more suggestion to try out on the team...

```
config zerotier 'holidays'
        option enabled '1'
        list join 'abcdef',allowDns='1',allowDefault='1'
        list join 'somethingElse'
```
Then a modified join(), and this is just a proof of concept - no defensive code at this time...
```

        add_join() {
                # an (empty) config file will cause ZT to join a network

                ZTnetwork=$(echo "$1" | cut -d ',' -f 1 )
                touch $path/networks.d/$ZTnetwork.conf

                echo "$1" | cut -d ',' -f 2 >> $path/networks.d/$ZTnetwork.local.conf
                echo "$1" | cut -d ',' -f 3 >> $path/networks.d/$ZTnetwork.local.conf
                echo "$1" | cut -d ',' -f 4 >> $path/networks.d/$ZTnetwork.local.conf
                echo "$1" | cut -d ',' -f 5 >> $path/networks.d/$ZTnetwork.local.conf

        }

        config_list_foreach $cfg 'join' add_join
```

This is what shows up in /var/lib/zerotier-one/networks.d/abcdef.local.conf

cat  /var/lib/zerotier-one/networks.d/9e1948db63701dd6.local.conf
```
allowDns=1
allowDefault=1


```
It would take a bit of more work to make it robust.  Just let me know what you guys think!

Comment (generalinterest) [reply zerotier/files/etc/init.d/zerotier]: Hi @mwarning 

I am just wondering if there was any thing further I could add to this issue?

I have updated the pull request to parse the 'list join' for the specified parameters, anything else will be ignored.

```
config zerotier 'holidays'
        option enabled '1'
        list join 'abcdef',allowDns='1',allowDefault='1'
        list join 'somethingElse'
```


```

add_join() {
		# an (empty) config file will cause ZT to join a network
		touch $path/networks.d/$1.conf

		ZTnetwork=$(echo "$1" | cut -d ',' -f 1 )
		PARAMS=$( echo "$1" | cut -d ',' -f 2-6)

		touch $path/networks.d/$ZTnetwork.conf

		# split the paramters into individual lines for easier parsing.
		# we just need the value of what the specific paramter equels to.
		allow_managed=$( echo "$PARAMS" | tr ',' '\n' | grep -w "allowManaged" | sed 's/^.*=//' )
		allow_glabal=$( echo "$PARAMS" | tr ',' '\n' | grep -w "allowGlobal" | sed 's/^.*=//' )
		allow_default=$( echo "$PARAMS" | tr ',' '\n' | grep -w "allowDefault" | sed 's/^.*=//' )
		allow_dns=$( echo "$PARAMS" | tr ',' '\n' | grep -w "allowDNS" | sed 's/^.*=//' )

                # create a corresponding .local.conf file if any of it's settings exist.
                if [ $allow_managed ] && [ -n $allow_managed ]; then
                        echo "allowManaged=$allow_managed" >> $path/networks.d/$ZTnetwork.local.conf
                fi
                if [ $allow_global ] && [ -n allow_managed ]; then
                        echo "allowGlobal=$allow_global" >> $path/networks.d/$ZTnetwork.local.conf
                fi
                if [ $allow_default ] && [ allow_default ]; then
                        echo "allowDefault=$allow_default" >> $path/networks.d/$ZTnetwork.local.conf
                fi
                if [ $allow_dns ] && [ -n allowd_dns ]; then
                        echo "allowDNS=$allow_dns" >> $path/networks.d/$ZTnetwork.local.conf
                fi
	}

	config_list_foreach $cfg 'join' add_join
```

What do you think? I would love to hear from the team about any feedback.

**Comment by mwarning** [comment]
Sorry, I di not have much time. The code looks like there should be a more elegant approach.

File: zerotier/files/etc/init.d/zerotier (hunk 1 above)
Comment (mwarning) [reply zerotier/files/etc/init.d/zerotier]: > @mwarning I haven't commented it out, but if it's OK with you I can try to rewrite `/etc/init.d/zerotier` to fit this configuration. 😉

I am fine with that approach. Sorry for my slow reaction.

**Comment by generalinterest** [comment]
Hi @mwarning 

How about this, where add_join calls a new function add_join_local.  This is named to indicate the parameters come from the /etc/config/zerotier "join" statement and destined for it's local.conf file.

The add_join_local takes two parameters, the full string of the join statement, and the second for the string to find and retrieve it's value.
 
The first parameter on the join is the network id, so we skip that and take all remaining fields.  Then the remaining string is split on a comma delimiter into separate lines.  It is then easy to find the string we are looking for and retrieve it's value as everything after the = sign.

It's then just a matter of appending that to the network.local.conf file where zerotier-one will pick it up.

What do you think?

```
        add_join_local() {

                # split the parameters into individual lines for easier parsing.
                # we just need the value of what the specific parameter equels to.

                local params=$( echo "$1" | cut -d ',' -f 2-)

                local allow_val=$( echo "$params" | tr ',' '\n' | grep -w "$2" | sed 's/^.*=//' )
                if [ $allow_val ] && [ -n allow_val ]; then
                        echo "$2=$allow_val" >> $path/networks.d/$ZTnetwork.local.conf
                fi
        }

        add_join() {
                # an (empty) config file will cause ZT to join a network

                local ZTnetwork=$(echo "$1" | cut -d ',' -f 1 )

                touch $path/networks.d/$ZTnetwork.conf

                add_join_local $1 "allowManaged"
                add_join_local $1 "allowGlobal"
                add_join_local $1 "allowDefault"
                add_join_local $1 "allowDNS"
        }
```

All the best
Brian

**Comment by ogarcia** [comment]
@mwarning since we have merged #120 I think we can close this PR. :wink:
